### ISSUE TITLE
Implement BASIL

### ISSUE THREAD

**Issue by HenkMutsaerts** [ExploreASL/ExploreASL#20]
# The following steps are needed as pre-work for BASIL
1. BASIL prefers masks to limit the voxels on which to run the modeling, but ExploreASL currently masks after the quantification. The masking wrapper should therefore be moved to before the quantification wrapper
    1. CBF will be replaced by PWI in the xASL_wrp_CreateAnalysisMask, after which this part can be moved in front of the Quantification
2. For better modeling BASIL requires the full 4D timeseries. Currently the resample wrapper does the averaging, unless there is a PWI4D option that states to keep the 4D timeseries. Instead, the averaging should move from the resample wrapper to the quantification wrapper, to ensure that the quantification wrapper always has access to the full 4D timeseries. The control-label subtraction can stay in the resample wrapper. The PWI4D flag can stay, but would be applied in the quantification wrapper only, where it would disable deleting of the temporary time series (this is for advanced users that want to keep the 4D time series for modelling purposes).
    1. Keep subtraction in xASL_wrp_Resample
    2. Move averaging to xASL_wrp_Quantify
    3. Remove PWI4D option from xASL_wrp_Resample
    4. Keep PWI4D option in xASL_wrp_Quantify

**Comment by HenkMutsaerts** [comment]
See also discussion with Martin Craig [HERE](https://docs.google.com/document/d/13QfpVda2JB0kxRAUdZRQbrMiuG7n9LB7q6lncVIPeF0/edit?usp=sharing). 
Thanks @ Martin Craig *martin.craig@eng.ox.ac.uk*

**Comment by HenkMutsaerts** [comment]
## 1 - Switch order quantification & mask creation
* xASL_module_ASL: switch order xASL_wrp_Quantify (now last) & xASL_wrp_CreateAnalysisMask (now first)
* Move creation of standard space masked image from xASL_wrp_CreateAnalysisMask to xASL_wrp_Quantify

## 1 - Replace CBF by PWI in xASL_wrp_CreateAnalysisMask (as CBF is not yet quantified now)
* xASL_wrp_CreateAnalysisMask -> replace 'CBF' by 'PWI'
* xASL_im_MaskNegativeVascularSignal -> replace 'CBF' by 'PWI'
* xASL_im_MaskPeakVascularSignal -> replace 'CBF' input by 'PWI' input
* Cosmetic: use 1. instead of 1)

**Comment by HenkMutsaerts** [comment]
When merging/rebasing edits from hotfix 1.2.1, note to change naming in:
xASL_im_MaskPeakVascularSignal:
![Screenshot 2020-08-20 at 14 39 16](https://user-images.githubusercontent.com/27774254/90773445-32c7ab80-e2f6-11ea-8317-99ff5c5a264e.png)

**Comment by HenkMutsaerts** [comment]
Successfully restored this branch by recreating it from new develop & cherry-picking its commits. Now needs to be tested & a new PR created.

**Comment by HenkMutsaerts** [comment]
I have now finished the pipeline preparations for BASIL implementation and tested these changes with `xASL_qc_TestExploreASL` on macOS Matlab 2016b. With thanks to martin.craig@eng.ox.ac.uk, check our discussion [HERE](https://docs.google.com/document/d/13QfpVda2JB0kxRAUdZRQbrMiuG7n9LB7q6lncVIPeF0/edit?usp=sharing). 

### Changes
>1. `070_Quantification` & `080_CreateAnalysisMask` are now switched to: `070_CreateAnalysisMask` & `080_Quantification`.

* BASIL needs a mask before quantification, to avoid running the full model on each voxel. 
* This order change should not affect the reproducibility.
* This order reversal is now appropriately tested and all dependencies fixed.
* In the respective masking functions, the CBF variables are replaced by PWI variables (as CBF is created only in the subsequent module)
* `xASL_wrp_Quantify`: now has an additional final step to create standard space `qCBF_masked`
* Definitions of `qCBF_masked`, MaskVascular and MaskSusceptibility are now moved to `xASL_init_FileSystem`

>2. xASL_wrp_Resample now always saves both PWI4D and PWI.

* The previous situation was to only save PWI4D if this was requested by `x.SavePWI4D`. 
* This option is now removed and PWI4D is always saved, both in native and standard space. These files are removed at the end of `xASL_wrp_VisualQC_ASL`, as this function can use PWI4D for QC parameters.
* `x.SaveCBF4D` is created, to allow keeping quantified time series of paired subtractions.

>3. While at it, I revamped `xASL_wrp_Resample`
* Layout is now improved for readability
* Resampling of motion correction now has 4th-degree B-spline resampling. This is the same as the warping to standard space and is the default of SPM. We previously used a 2nd-degree B-spline here, which is a good compromise for images that are not used for quantification (e.g. images for registration).
* Bilateral filter is now disabled as this needs testing


>4. While at it, I revamped some other minor things:

* `xASL_adm_LoadParms`: `x.M0=no_background_suppression` now issues a warning (should be `x.M0=UseControlAsM0`)
* `xASL_im_Presmooth`: now loads image matrices as single precision instead of double, perhaps this speeds up the `convn` function
* `xASL_im_Upsample`: now issues an error when input image has too many dimensions
* `xASL_qc_CreatePDF`: minor bugfix to allow printing booleans

**Comment by HenkMutsaerts** [comment]
Seems this issue can be closed for 1.4.0, and perhaps reopened for 1.5.0

**Comment by jan-petr** [comment]
> Seems this issue can be closed for 1.4.0, and perhaps reopened for 1.5.0

I agree. I will run the test data. Then, during the merge, conflicts need to be resolved, but these are uniquely in the function headers, so I will do that.

**Comment by jan-petr** [comment]
This issue has been merged and branch deleted. And will be released in 1.4.0 for further work on implementing BASIL in ExploreASL, @HenkMutsaerts needs to create a new issue!!!

**Comment by HenkMutsaerts** [comment]
Sure, but we can re-open this issue in 1.5.0 right?
@jan-petr

**Comment by HenkMutsaerts** [comment]
Answers below inline (applying to the latest version on the develop branch (git checkout develop))
 
>> I was originally applying the mask to Basil by taking voxels in the PWI that are not NaN - since I thought the mask generation function sets masked voxels to NaN. But it doesn't look like this is right as it ends up with almost all the volume included. How can I access the boolean analysis mask from within the quantification function? (Sorry couldn't figure it out!)
 

You’re right, we deliberately keep the mask and the image separate, which allows visualizing the original images also without masking in any space, which can be useful for QC.
 
The analysis mask (as explained in the ExploreASL manuscript) is separated in 3 masks:
	•	1) FoV mask (including voxels that were scanned or not scanned, but voxels outside the FoV should be set to NaN in any space;
	•	2) vascular mask, removing extreme values (at negative or positive side). However, you may not want to apply this mask for quantification that uses spatial averaging (that you may use in multi-TI?);
	•	3) susceptibility artifact masking based on probability masks (in standard space only)
 

I would start with a brainmask, when putting a breakpoint/debugging in quantification wrapper:
% BrainMask
`Path_pvGM = fullfile(x.P.SessionDir, 'PVgm.nii');`
`Path_pvWM = fullfile(x.P.SessionDir, 'PVwm.nii');`
`pvWholeBrain = xASL_io_Nifti2Im(Path_pvGM) + xASL_io_Nifti2Im(Path_pvWM);`
`MaskWholeBrain = pvWholeBrain>0.5;`
 
% If you then want to remove the vascular extremes as well, you could add:
`MaskVascular = logical(xASL_io_Nifti2Im(x.P.Path_MaskVascular));`
`Mask = MaskWholeBrain & MaskVascular;`
 
>> Do I need to do anything extra regarding 4D quantification? Currently if SavePWI4D is defined then in addition to the averaged volume, Basil also gets called on the full 4D volume. Is the output from the two separate runs kept separately or will the 4D processing overwrite the other?
 

Per #20, I removed option x.SavePWI4D; and a PWI4D.nii is now always saved. So the quantification module will always need to do the averaging as well, and saved to a single 3D volume CBF.nii. If x.Q.SaveCBF4D == true, then the individual subtractions should be quantified as well and stored as CBF4D.nii.
(This is reflected in xASL_module_ASL line 419: if isfield(x.Q, 'SaveCBF4D') && x.Q.SaveCBF4D==1
 
>> I also had a problem with  Functions/xASL_import_json.m  
The code which checks the Matlab version didn't work for me because my version string is …
 

You’re right, this is being fixed, sorry.
 
>> Finally, Basil is mainly useful when you have multi-delay data - I wasn't clear whether ExploreASL supports this yet - if so do you have a test data set for this?
 

No, we don’t support multi-delay yet, this is indeed one of the main reasons to call BASIL 😊
The multi-PLD example that we will work with, is actually your/BASILs example, which is online as BIDS example but I need to transform it to the ExploreASL format still. So you can either wait on this, or it is probably easiest if you take your own example?
 
e.g.: https://github.com/patsycle/bids-examples/tree/BEP005/ASL_Siemens_PCASL_multiPLD_pepolar
e.g.2: https://github.com/patsycle/bids-examples/tree/BEP005/ASL_Siemens_PASL_multiTI

**Comment by HenkMutsaerts** [comment]
Apologies I incorrectly reopened this.
